auth/authenticate_with_aws_provider.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package snippets
+
+// [START auth_cloud_aws_provider]
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"cloud.google.com/go/auth/credentials/externalaccount"
+	"cloud.google.com/go/storage"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+)
+
+type customAwsProvider struct {
+	awsRegion string
+}
+
+func (acp customAwsProvider) AwsRegion(ctx context.Context, opts *externalaccount.RequestOptions) (string, error) {
+	return acp.awsRegion, nil
+}
+
+func (acp customAwsProvider) AwsSecurityCredentials(ctx context.Context, opts *externalaccount.RequestOptions) (*externalaccount.AwsSecurityCredentials, error) {
+	// Load the AWS default config and retrieve the credentials.
+	cfg, err := config.LoadDefaultConfig(ctx)
+	if err != nil {
+		return nil, err
+	}
+	val, err := cfg.Credentials.Retrieve(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert the AWS credentials into the external account libraries version of AWS security credentials.
+	awsCredentials := &externalaccount.AwsSecurityCredentials{
+		AccessKeyID:     val.AccessKeyID,
+		SecretAccessKey: val.SecretAccessKey,
+		SessionToken:    val.SessionToken,
+	}
+
+	return awsCredentials, nil
+}
+
+// authenticateExplicitWithAdc uses Application Default Credentials
+// to print storage buckets.
+func authenticateExplicitWithAwsProvider(w io.Writer) error {
+	ctx := context.Background()
+
+	// Set the active AWS region, i.e. "us-east-2"
+	awsRegion := "us-east-2"
+	awsCredentialsProvider := customAwsProvider{awsRegion: awsRegion}
```

Comment:
issue: pass this in as an anonymous struct.

Reply:
It seems like implementing an interface (which is required here) with an anonymous struct is a bit messy, would look something like this unless I'm missing something:

```
awsCredentialsProvider := struct {
    region string // Could also get rid of this and have the struct be completely empty.
}{
    region: "us-east-2",
}

func (provider struct {
    region string
}) AwsRegion(ctx context.Context, *opts RequestOptions) (string, error) {
    return region, nil
}

func (provider struct {
    region string
}) AwsSecurityCredentials(ctx context.Context, *opts RequestOptions) (AwsSecurityCredentials, error) {
  // Logic here
}

```
It doesn't seem to me like making the struct anonymous makes the code any simpler or clearer, and IMO makes it a little less obvious that a interface is being implemented. Thoughts? This seems like an example where having a helper struct/functions makes sense since in general we would expect customers to create a struct that implements the Aws credentials provider rather than use an anonymous struct.

Reply:
Couldn't you just do this?

```
	awsCredentialsProvider := struct {
		awsRegion string
	}{
		awsRegion: "us-east2",
	}
```

Reply:
Yeah that would create an anonymous struct, but the struct needs to implement the AwsSecurityCredentialsProvider interface to be used to create a new external account struct (line 80) by having a defintion for the AwsRegion and AwsSecurityCredentials functions. If I remove the customAwsProvider struct and change it to be anonymous, then I have to have functions that act on an anonymous struct as well, leading to needing two functions set up like this

```
func (provider struct {
    region string
}) AwsRegion(...) (string, error) {
    return region, nil
}
```

vs having the explicit struct which would allow

```
func(c customAwsProvider) AwsRegion (...) (string, error) {
   return c.region, nil
}
```

the helper struct and functions are there to be the most basic working implementation of the interface that we would expect users to add on to, apologies if that wasn't clear, probably need to add comments.

Reply:
Oh, I didn't catch the pointer receivers in those two functions. I see.

Reply:
I think I would still move this struct and its methods inside of the sample function.